src/main/scala/com/gu/identity/testing/usernames/TestUsernames.scala

```diff
@@ -1,63 +1,89 @@
+
 import java.nio.ByteBuffer
 import java.time.Clock.systemUTC
 import java.time.{Clock, Duration, Instant}
 import java.time.Duration.ofMinutes
 import java.time.Instant.ofEpochSecond
-
 import com.typesafe.scalalogging.LazyLogging
 
 
 
-trait TestUsernames {
-  def generate(conf: Array[Byte]): String
+class TestUsernames(usernameEncoder: Encoder, recency: Duration)(implicit clock: Clock) extends LazyLogging {
+  def generate(conf: Array[Byte]): String = {
+    val bb = ByteBuffer.allocate(Encoder.PayloadByteLength)
+    bb.putInt(Instant.now(clock).getEpochSecond.toInt)
+    bb.put(conf)
+    usernameEncoder.encodeSigned(bb.array())
+  }
 
-  def validate(username: String): Option[Array[Byte]]
+  def validate(username: String): Option[Array[Byte]] = {
 
-  def generate(): String = generate(Array.ofDim[Byte](2))
+    import scala.math.Ordering.Implicits._
 
-  def isValid(username: String): Boolean = validate(username).isDefined
-}
-
-object TestUsernames extends LazyLogging {
+    def extractRecentConfData(validlySignedData: Array[Byte]): Option[Array[Byte]] = {
+      val bb = ByteBuffer.wrap(validlySignedData)
+      val creationTime = ofEpochSecond(bb.getInt())
+      val now = Instant.now(clock)
+      if (creationTime > now) {
+        logger.warn(s"TestUsername created on $creationTime, apparently AFTER this moment in time! now=$now")
+      }
 
-  val ConfPayloadByteSize = 2
+      val confPayload: Array[Byte] = Array.ofDim[Byte](ConfPayloadByteSize)
+      bb.get(confPayload)
 
-  def apply(usernameEncoder: Encoder, recency: Duration = ofMinutes(30))(implicit clock: Clock = systemUTC): TestUsernames = new TestUsernames {
-    def generate(conf: Array[Byte]): String = {
-      val bb = ByteBuffer.allocate(Encoder.PayloadByteLength)
-      bb.putInt(Instant.now(clock).getEpochSecond.toInt)
-      bb.put(conf)
-      usernameEncoder.encodeSigned(bb.array())
+      val expirationTime = creationTime.plus(recency)
+      if (expirationTime > now) Some(confPayload) else {
+        logger.warn(s"TestUsername created on $creationTime EXPIRED on $expirationTime")
+        None
+      }
     }
 
-    def validate(username: String): Option[Array[Byte]] = {
+    for {
+      validlySignedData <- usernameEncoder.decodeSigned(username)
+      recentlyCreatedConfData <- extractRecentConfData(validlySignedData)
+    } yield recentlyCreatedConfData
+  }
+
+  def isValid(username: String): Boolean = validate(username).isDefined
+
+  def generateEmail(baseEmail: String): TestUser = {
+    val token = generate(Array.ofDim[Byte](2))
+    val email: String = tokenToEmail(baseEmail, token)
+    TestUser(email, token)
+  }
 
-      import scala.math.Ordering.Implicits._
+  def isValidEmail(primaryEmailAddress: String): Boolean = {
+    val maybeValidTestUser = maybeTokenFromEmail(primaryEmailAddress)
+    maybeValidTestUser.exists(isValid)
+  }
```

Comment:
these are the two new functions that call `tokenToEmail` and `maybeTokenFromEmail` in order to have a consistent idea of what test user emails looks like